squad-get-log-lines: Create log fetching script

katieworton opened the pull request:
Add a script which fetches the logs for a specified group, project list, and suite list over `--count` number of builds for each group/project combination.

This script will fetch the log parameter for the tests within the group/project/suite combination specified. The logs are placed in the following paths:
`logs/<test.name>/<project.slug>/<test.id>`

The script can be run by a command of the following format: 
```
python squad-get-log-lines --project linux-next-master --group lkft --suite log-parser-boot log-parser-test --debug
```

katieworton committed:
squad-get-log-lines: Create log fetching script

Add a script which fetches the logs for a specified group, project list,
and suite list over `--count` number of builds for each group/project
combination.

This script will fetch the log parameter for the tests within the
group/project/suite combination specified. The logs are placed in the
following paths:
`logs/tests/<test.name>/<project.slug>/<test.id>`

The script can be run by a command of the following format:
python squad-get-log-lines --group lkft --projects linux-next-master \
linux-mainline-master linux-stable-rc-linux-5.15.y \
linux-stable-rc-linux-5.10.y linux-stable-rc-linux-6.1.y \
linux-stable-rc-linux-5.4.y linux-stable-rc-linux-6.6.y \
linux-stable-rc-linux-4.19.y --suites log-parser-boot log-parser-test \
--debug

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
This still needs an update to the README, but thought I would put this here as-is to get initial feedback.

chaws commented on squad-get-log-lines:
```diff
@@ -0,0 +1,107 @@
+        builds = project.builds(count=args.count)
+
+        for build_id in builds:
+            build = Build(build_id)
```
Doing `Build(build_id)` makes a new request to SQUAD, sorry I should've documented that somewhere :/ 

What you can do in the line above is
```python
for build_id, build in builds.items():
```

then you wouldn't need `build = Build(build_id)`

katieworton replied:
That's a great spot - thank you for the feedback! I've made this change in both places :)

chaws commented on squad-get-log-lines:
```diff
@@ -0,0 +1,107 @@
+            # Look at all the tests in the build
+            tests = build.tests(count=ALL, **filters)
+
+            for test_id in tests:
```
Same here...

katieworton commented:
@roxell If you could have a look over this script that would be great! In particular, it would be helpful  if you could check:
- Is the directory structure and file naming produced sensible and at the correct granularity for checking logs across different test names and projects?
- Are the script inputs sensible and easy enough to work with?

Any other feedback also welcome - those were just two things in particular I wasn't sure about when drafting this :)

roxell commented on squad-get-log-lines:
```diff
@@ -0,0 +1,105 @@
+    )
+
+    parser.add_argument(
+        "--projects",
```
In the changelog it says "--project linux-next-master"

katieworton replied:
fixed :)

roxell commented on squad-get-log-lines:
```diff
@@ -0,0 +1,105 @@
+    )
+
+    parser.add_argument(
+        "--suites",
```
In the changelog it says "--suite log-parser-boot log-parser-test"

katieworton replied:
fixed :raised_hands:

roxell commented on squad-get-log-lines:
```diff
@@ -0,0 +1,105 @@
+                            Path("logs")
+                            / Path(test.name)
+                            / Path(project.slug)
+                            / Path(str(test.id))
```
maybe the test_id should be a directory and store a file called logline and then you have the option to store the logfiles per testrun somewhere and create a symlink to that logfile inside the test_id directory.
something like
logs/test.name/project.slug/test.id/logline
logs/logfiles/testrun.id.log
ln -s logs/logfiles/testrun.id.log logs/test.name/project.slug/test.id/logfile

katieworton replied:
I've added a few more patches to this PR (one patch to add support for log file download and a couple to deal with authentication headers for the file downloads).

To summarise, I've:
- Added support for downloading the full logs
- When full logs are downloaded, a directory structure is created which creates symlinks to the full logs next to symlinks to all the log lines which came from that full log
- Tweaked the initial patch slightly so it aligns a bit better with the above changes

I thought this way it would be easier to see the relationship between the original log and all the different log lines that came from that log.

Let me know what you think :+1:

katieworton committed:
squad-get-log-lines: Add flag to download full logs

Add a `--full-log` flag to support downloading the full logs alongside
the log entries for tests.

When this `--full-log` flag is provided, it will download the full log
for each failing test to the following paths:

`logs/testruns/<test.name>/<project.slug>/<testrun.id>/full_log_<testrun.id>`

It will also create a directory which places together symbolic links to
the full log and its associated test log lines. This makes it easier to
see what log lines were extracted from each full log. These symlinks are
placed in the following paths:

`logs/testruns_next_to_tests/<project.slug>/<testrun.id>/`

With the full log symlink being stored in a file `full_log_<testrun.id>`

And the test logs symlinks being stored in `<suite>/<test.id>`

An example of how to run with this flag is:
python squad-get-log-lines --group lkft --projects linux-next-master \
linux-mainline-master linux-stable-rc-linux-5.15.y \
linux-stable-rc-linux-5.10.y linux-stable-rc-linux-6.1.y \
linux-stable-rc-linux-5.4.y linux-stable-rc-linux-6.6.y \
linux-stable-rc-linux-4.19.y --suites log-parser-boot \
log-parser-test --test-count 150 --debug --full-log

Note: the test-count has been limited to 150 in this example to prevent
an excessive number of logs from being downloaded.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: Add support for request headers to get_file

Add support for request headers to get_file to allow authorization
headers to be used in the get requests for downloads.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-get-log-lines: Add SQUAD_TOKEN header to log file download

Add the SQUAD_TOKEN authorization header to the get_file call to make
requests to download the full logs authenticated.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

roxell commented on squad-get-log-lines:
```diff
@@ -0,0 +1,199 @@
+        action="store_true",
+        default=False,
+        help="Download the full log in addition to the log lines.",
+    )
```
Can this be default True ?
We don't really need to pull out any boot/test-log-parser tests lines.
Since that can be done if we just pull out the logfiles and run the logparser locally once.
Then change the regexes and rerun it locally, now you have two different runs of the logparser run that can be compared and see if regex changes was an improvement or not before committing it.

The script will be way simpler if we drop the boot/test-log-parser tests and only pull the complete logfiles.
